### PR TITLE
Support `BigInteger` and `BigDecimal` literals in SpEL expressions

### PR DESCRIPTION
**Pull request by ovmarkushyn**
### Context

The change satisfies the request made in #21758 which involves adding support for a suffix to `BigInteger`, similar to 'L' for `long`.

### Usage

The proposed solution makes it possible to use both `BigInteger` and `BigDecimal` for both HEX and non-HEX number literals.

The suffixes `BI` and `BD` are used for `BigInteger` and `BigDecimal`, respectively. Both suffixes can be specified in lower/upper case or mixed.

### Solution

Consider please diagram attached.

It shows all the elements that have been changed as part of the solution.

**Yellow** - represent existing elements.

**Green** - "big number" (`BigInteger` and `BigDecimal`) extension of `Tokenizer`.

In general, the solution is implemented similarly to the `L` support, aiming for as much consistency as possible with the current design.

Whole "big number" related code is isolated in 2 inner classes, one for **HEX** literal (Big**Hex**IntegerConcernExtension) and another one for **non-HEX** one (Big**NonHex**NumberConcernExtension) both having common parent `BigNumberConcernExtension`.

`BigNumberConcernExtension` keeps all the logic related to suffix detection, extraction, position shifting.

`BigHexIntegerConcernExtension` and `BigNonHexNumberConcernExtension` are responsible for handling the specifics of token addition.

[](https://github.com/spring-projects/spring-framework/assets/1871503/6ec2f923-933e-472d-a6b0-01d6477d9556)

### CODE REVIEW

File: spring-expression/src/main/java/org/springframework/expression/spel/BigNumberConcern.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * Marker of "big number" ({@link java.math.BigInteger} and {@link java.math.BigDecimal}) concern extension for SpEL engine.
+ * Solution extends related elements similar to how L/l suffix support is done.
+ * It covers: HEX, non-HEX(including exponent) literals.
+ * Requirement: <a href="https://github.com/spring-projects/spring-framework/issues/21758">Suffix support for BigInteger in SPEL [SPR-17225] #21758</a>
+ *
+ * @author Oleksandr Markushyn
+ */
+public @interface BigNumberConcern {
```
Comment (sbrannen) [inline spring-expression/src/main/java/org/springframework/expression/spel/BigNumberConcern.java:27]: Please remove the `@BigNumberConcern` from this PR.

We do not use marker annotations in Spring Framework to track features, and its presence in this PR makes it harder to review the actual changes in the PR.

**Review by sbrannen — CHANGES_REQUESTED** [review]

File: spring-expression/src/main/java/org/springframework/expression/spel/BigNumberConcern.java (hunk 1 above)
Comment (ovmarkushyn) [reply spring-expression/src/main/java/org/springframework/expression/spel/BigNumberConcern.java]: Removed.

**Comment by ovmarkushyn** [comment]
@sbrannen 
BigNumberConcern is also removed from PR description and diagram.

**Comment by sbrannen** [comment]
Hi @ovmarkushyn,

In light of https://github.com/spring-projects/spring-framework/issues/21758#issuecomment-1937763112, we are closing this PR as well.

Having said that, we do appreciate the hard work you put into this PR! 👍 

Thanks,

Sam

**Pull request closed without merging**